### DISCUSSION TITLE
CallPeaks on a merged Seurat Object

### DISCUSSION THREAD

**Discussion by shinlin77** [stuart-lab/signac#854] [Q&A]
I'm trying to perform CallPeaks on a merged Seurat Object.  The cell barcodes are renamed in the merged object because 2 separate runs may use the same barcode name.   When I run CallPeaks on the merged object, it appears that no cell barcodes are being picked up (i.e. "Keeping 0 cell barcodes")--I'm assuming because the names in the fragment files are different from the renamed cell names in the merged file.  Later, it appears macs2 has nothing to work on (therefore, the division by zeros).  Anyway, that's my interpretation of the output.

Is there a fix for this?  Thanks.

\>peaks  = CallPeaks(multiome_data, macs2.path = "/scg/apps/legacy/software/macs/2.1.1/bin/macs2", group.by = "cell_type")

Processing file /oak/SREQ-315_5_results/outs/atac_fragments.tsv.gz
Keeping 0 cell barcodes
Splitting into 12 files
Done Processing 145 million lines

Processing file /oak/SREQ-315_6_results/outs/outs/atac_fragments.tsv.gz
Keeping 0 cell barcodes
Splitting into 12 files
Done Processing 170 million lines

INFO  @ Wed, 03 Nov 2021 20:24:55:
\# Command line: callpeak -t /local/scratch/slrmtmp.28499376/RtmpYVxQyO/Fibroblast.bed -g 2.7e+09 -f BED --nomodel --extsize 200 --
shift -100 -n Fibroblast --outdir /local/scratch/slrmtmp.28499376/RtmpYVxQyO
\# ARGUMENTS LIST:
\# name = Fibroblast
\# format = BED
\# ChIP-seq file = ['/local/scratch/slrmtmp.28499376/RtmpYVxQyO/Fibroblast.bed']
\# control file = None
\# effective genome size = 2.70e+09
\# band width = 300
\# model fold = [5, 50]
\# qvalue cutoff = 5.00e-02
\# Larger dataset will be scaled towards smaller dataset.
\# Range for calculating regional lambda is: 10000 bps
\# Broad region calling is off
\# Paired-End mode is off

INFO  @ Wed, 03 Nov 2021 20:24:55: #1 read tag files...
INFO  @ Wed, 03 Nov 2021 20:24:55: #1 read treatment tags...
Exception ZeroDivisionError: 'integer division or modulo by zero' in 'MACS2.IO.Parser.GenericParser.tsize' ignored
INFO  @ Wed, 03 Nov 2021 20:24:55: #1 tag size is determined as 0 bps
INFO  @ Wed, 03 Nov 2021 20:24:55: #1 tag size = 0
INFO  @ Wed, 03 Nov 2021 20:24:55: #1  total tags in treatment: 0
INFO  @ Wed, 03 Nov 2021 20:24:55: #1 user defined the maximum tags...
INFO  @ Wed, 03 Nov 2021 20:24:55: #1 filter out redundant tags at the same location and the same strand by allowing at most 1 tag(s)
INFO  @ Wed, 03 Nov 2021 20:24:55: #1  tags after filtering in treatment: 0
Traceback (most recent call last):
  File "/scg/apps/legacy/software/macs/2.1.1/bin/macs2", line 4, in <module>
    __import__('pkg_resources').run_script('MACS2==2.1.1.20160309', 'macs2')
  File "build/bdist.linux-x86_64/egg/pkg_resources/__init__.py", line 719, in run_script
  File "build/bdist.linux-x86_64/egg/pkg_resources/__init__.py", line 1498, in run_script
  File "/scg/apps/legacy/software/macs/2.1.1/lib/python2.7/site-packages/MACS2-2.1.1.20160309-py2.7-linux-x86_64.egg/EGG-INFO/scripts/macs
2", line 617, in <module>

  File "/scg/apps/legacy/software/macs/2.1.1/lib/python2.7/site-packages/MACS2-2.1.1.20160309-py2.7-linux-x86_64.egg/EGG-INFO/scripts/macs
2", line 57, in main

  File "build/bdist.linux-x86_64/egg/MACS2/callpeak_cmd.py", line 112, in run
ZeroDivisionError: float division by zero
Error in file(file, "rt") : cannot open the connection
In addition: Warning message:
In file(file, "rt") :
  cannot open file '/local/scratch/slrmtmp.28499376/RtmpYVxQyO/Fibroblast_peaks.narrowPeak': No such file or directory

**Comment by timoast** [comment]
How did you construct the Seurat object?

**Comment by shinlin77** [comment]
counts = Read10X_h5(...
object = CreateSeuratObject(counts = counts$'Gene Expression', assay =
"RNA")
object[["ATAC"]] = CreateChromatinAssay(counts = counts$Peaks...

I think I've found a way around this by completing renaming all the cells
(even within the atac_fragments.tsv.gz  files).  Maybe there is a more
elegant solution to this problem.

**Reply by timoast** [reply]
You shouldn't need to rename cells in the fragment files, we have ways to handle this in Signac. If you provide the code I can help more

**Reply by yjchen1201** [reply]
Hi Tim, 

To follow up, I also got the same error as @shinlin77 's. My code is pasted below. 
I have 10 samples and I merged them together using the common peak set. Then I performed the basic workflow to identify clusters. But when I tried calling cluster-specific peaks using CallPeaks function, the error happened. I checked the splitted cluster fragment bed files in my "frag_temdir" directory, I found that some cluster's fragment bed files were missing. So i guess that probably caused the final error.  Could you please help with solving this issue? Or Could there be something wrong in my steps?
Thank you!  Love this amazing tool!

```{r}
# Example of loading seurat object created using cellrangerARC filtered matrix, 10 samples 
obj_1<-readRDS("obj_1_seurat_ATAC.rds")
# Example of loading called peaks using each sample as a pseudo bulk, 10 peak files 
peak.obj_1<-readRDS("obj_1_macs2peaks.rds") #GRanges object 

# Combined called peaks from 10 samples 
combined.peaks <- reduce(x = c(peaks.obj_1, peaks.obj_2, peaks.obj_3, peaks.obj_4, peaks.obj_5, peaks.obj_6,peaks.obj_7, peaks.obj_8,peaks.obj_9, peaks.obj_10))

# Filter out bad peaks based on length
peakwidths <- width(combined.peaks)
combined.peaks <- combined.peaks[peakwidths  < 10000 & peakwidths > 20] 

# Example of creating fragment object, 10 samples
frags.obj_1 <- CreateFragmentObject(
  path = "obj_1/outs/atac_fragments.tsv.gz",
  cells = colnames(obj_1)
)
# Example of create count matrix
counts.obj_1 <- FeatureMatrix(
  fragments = frags.obj_1,
  features = combined.peaks,
  cells = colnames(obj_1)
)
assay.obj_1 <- CreateChromatinAssay(counts.obj_1, fragments = frags.obj_1)
obj_1 <- CreateSeuratObject(assay.obj_1 , assay = "ATAC_macs2")

# Rename cells before merging by adding prefix
obj_1<-RenameCells(obj_1, new.names = paste('obj1', colnames(x =  obj_1), sep = '_'))
#### Merge objects
merged.atac <- merge(obj_1, y = c(obj_1,obj_2,obj_3,obj_4,obj_5,obj_6,obj_7,obj_8,obj_9,obj_10),
    merge.data = TRUE)

sample="ATAC_merged"
#### process the combined dataset
merged.atac <- FindTopFeatures(merged.atac, min.cutoff = 10)
merged.atac <- RunTFIDF(merged.atac)
merged.atac <- RunSVD(merged.atac)
merged.atac <- FindNeighbors(merged.atac, verbose = TRUE, dims = 2:30,reduction = "lsi")
merged.atac <- FindClusters(merged.atac, verbose = TRUE, resolution = 0.6,,reduction = "lsi")
merged.atac <- RunUMAP(merged.atac, reduction = "lsi", dims = 2:30)

# Call peaks again, group.by clusters
assay="ATAC_macs2"
DefaultAssay(merged.atac) <- assay
working_path<-getwd()
macs2_output_path <- paste(working_path,"macs2_callpeak",sep="/")
dir.create(macs2_path)
frag_temdir <-paste(working_path,"macs2_callpeak","frag_tempdir",sep="/")
dir.create(frag_temdir)

macs2peaks<-CallPeaks(
  object=merged.atac,
  macs2.path = "/opt/apps/python2/bin/macs2",
  outdir = macs2_output_path,
  group.by="seurat_clusters",
  fragment.tempdir=frag_temdir,
  combine.peaks=T,
  extsize = 200,
  shift = -100,
  effective.genome.size = 1.4e+09,
  additional.args = "--bdg --SPMR -f BEDPE",
  name = sample,
  cleanup = FALSE,
  verbose = TRUE,
)
```
